Specify Arrow as an external dependency

jclay opened the pull request:
Do not pack Arrow source files into the final bundle, rely on them being provided externally. This should fix #213.

domoritz commented on src/mapd-con-es6.js:
```diff
@@ -23,7 +23,7 @@ import EventEmitter from "eventemitter3"
 import MapDClientV2 from "./mapd-client-v2"
 import processQueryResults from "./process-query-results"
-import { Table } from "apache-arrow"
+import * as Arrow from "apache-arrow"
```
Is this change necessary?

israelvicars commented:
@jclay Do we need to rebuild so that the arrow source files are excluded?

jclay commented:
@israelvicars good catch. I'm wondering what the current practice is with keeping built artifacts like this in the repo. Do you see a reason for keeping them checked in going forward?

israelvicars commented:
@jclay Yeah, that's been a matter of some debate. I've missed some of the previous repartee. My understanding is that pre-building is a practice with several deps to make them more accessible to a subset of users that would benefit from that. Dropping in a  That's rather vague and useless, just recalling echos of conversations... Because it's a practice across several repos, I think it merits more digging and consensus before we change practices. But we should change as quickly as possible if we can get some new consensus. I'd like to avoid doing it, personally.

domoritz commented:
`5.5.2-rc3` doesn’t seem to work for me. See https://observablehq.com/d/12c6ce12edb5f725

jclay commented:
@domoritz this should be resolved now. Are you able to give `5.5.2@rc4` a try? 

Observable's require was picking up the package as a `UMD` module, so when the cell was evaluated:

```
MapdCon = require("@mapd/connector@5.5.2-rc3").catch(() => window.MapdCon)
```

`catch` was never used. If we use `then` in place, it will properly override this to use the `MapdCon` defined on window.

I've now fixed some issues with our UMD packaging so that the following will work:

```
MapdCon = {
    const MapdCon = await require("@mapd/connector@5.5.2-rc4")
    return MapdCon.Connector
}
```

require of course returns a module, rather than the `MapdCon` class, so I've updated this to export `Connector` to clarify.

domoritz commented:
Even better, this works now in Observable (note the even simpler import):

```
MapdCon = require("@mapd/connector@5.5.2-rc4")

connector = new MapdCon.Connector()
      .protocol("https")
      .host("metis.mapd.com")
      .port("443")
      .dbName("mapd")
      .user("mapd")
      .password("HyperInteractive")
      .connectAsync()
```

or if you want the connector directly

```
MapdCon = (await require("@mapd/connector@5.5.2-rc4")).Connector
```

I wonder, could you make `Connector` the default export so that we may not need the indirection anymore?

domoritz commented:
This still does not seem to fix https://github.com/omnisci/mapd-connector/issues/213. 

<img width="832" alt="Screen Shot 2021-01-06 at 10 53 34" src="https://user-images.githubusercontent.com/589034/103755036-07eaa500-4fc2-11eb-9f48-c78176c3a155.png">

@jheer do you know a way to make https://github.com/vega/vega-loader-arrow/blob/36e1d0d2b0050f6e207327e78367007c4d25e28b/src/arrow.js#L20 more robust to different ways the arrow table is constructed?

jclay commented:
@domoritz OK, good to hear we're making some progress on this in Observable at least. I was hopeful that the external bundle would have solved #213. @israelvicars perhaps someone on the FE team would know a better way to tackle this.

jclay commented:
@domoritz The only way I've found to handle export default is with the following:

Change the entrypoint export:

```js
export { MapdCon as default } from "./mapd-con-es6";
```

However, Observable's use of d3-require seems to still need the following:

```js
const MapdCon = (await require("http://localhost:8080/browser-connector.js")).default;
```

I can't determine why it doesn't seem to handle the default export, well...by default. Am I missing something?

domoritz commented:
Can you do

```
import MapdCon from "./mapd-con-es6";
export = MapdCon;
```

bmatcuk commented:
I think I've figured this out, but I couldn't push my changes to this PR since @jclay's PR is based on a fork - I created my own PR here: https://github.com/omnisci/mapd-connector/pull/218

bmatcuk commented:
Closed in favor of #218
